### PR TITLE
Feature/interpreter/sensor detection range

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Types of PR

- [x] New Features
- [x] Upgrade of existing features
- [ ] Bugfix

### New features
- New feature of setting detection sensor range

### Ugrade of existion features
- change behavior based on the parameter described in the scenario file

Current implementation:

| detectionSensorRange [m] | isClairvoyant   | detectedObjects                                 |
| ------------------------ | --------------- | ----------------------------------------------- |
| 300 [m] (default)        | false (default) | Objects detected by LiDAR raycast in 100m range |
| 300 [m] (default)        | true            | All Objects within a 300m range                 |
| 10 [m]                   | false (default) | Objects detected by lidar raycast in 100m range |
| 10 [m]                   | true            | All Objects within a 10m range range             |

After this PR:

| detectionSensorRange [m] | isClairvoyant   | detectedObjects                                     | detectable objects |
| ------------------------ | --------------- | --------------------------------------------------- | ------------------ |
| 300 [m] (default)        | false (default) | Objects detected by lidar raycast in **300m** range | A                  |
| 300 [m] (default)        | true            | All Objects within a **300m** range                 | A, B, C            |
| 10 [m]                   | false (default) | Objects detected by lidar raycast in **10m** range   | -                  |
| 10 [m]                   | true            | All Objects within a **10m** range range             | B                  |

image

## How to review this PR.

I will prepare scenario file.

### CODE REVIEW

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/sensor_simulation/lidar/raycaster.hpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ class Raycaster
   }
   const sensor_msgs::msg::PointCloud2 raycast(
     std::string frame_id, const rclcpp::Time & stamp, geometry_msgs::msg::Pose origin,
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/include/simple_sensor_simulator/sensor_simulation/lidar/raycaster.hpp:52]:
```suggestion
    const std::string & frame_id, const rclcpp::Time & stamp, const geometry_msgs::msg::Pose & origin,
```

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/simulation_interface/proto/simulation_api_schema.proto [hunk 2]
```diff
@@ -46,7 +45,7 @@ message DetectionSensorConfiguration {
   double update_duration = 2;           // Update duration of the detection sensor. (unit : second)
   double range = 3;                     // Sensor detection range. (unit : meter)
   string architecture_type = 4;         // Autoware architecture type.
-  bool filter_by_range = 5;             // If false, simulator publish detection result only lidar ray was hit. If true, simulator publish detection result of entities in range.
+  bool detect_all_objects_in_range = 5;             // If false, simulator publish detection result only lidar ray was hit. If true, simulator publish detection result of entities in range.
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply simulation/simulation_interface/proto/simulation_api_schema.proto]: changed in https://github.com/tier4/scenario_simulator_v2/pull/1056/commits/4656077273ae3e24a4b794a7326251a1339d63b5

File: simulation/simulation_interface/proto/simulation_api_schema.proto [hunk 3]
```diff
@@ -64,7 +63,7 @@ message OccupancyGridSensorConfiguration {
   uint32 height = 5;            // Height of the occupancy grid. (unit : pixel)
   string architecture_type = 6; // Autoware architecture type.
   double range = 7;             // Sensor detection range. (unit : meter)
-  bool filter_by_range = 8;     // If false, simulator publish detection result only lidar ray was hit. If true, simulator publish detection result of entities in range.
+  bool detect_all_objects_in_range = 8;     // If false, simulator publish detection result only lidar ray was hit. If true, simulator publish detection result of entities in range.
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply simulation/simulation_interface/proto/simulation_api_schema.proto]: deleted this parameter in https://github.com/tier4/scenario_simulator_v2/pull/1056/commits/4656077273ae3e24a4b794a7326251a1339d63b5

**Comment by hakuturu583** [comment]
@kyoichi-sugahara Thank you for fixing problems, but the build test was failed.

**Comment by kyoichi-sugahara** [comment]
@hakuturu583 
I solved the failing of the build test. Thank you for checking!

File: simulation/simple_sensor_simulator/src/sensor_simulation/detection_sensor/detection_sensor.cpp [hunk 4]
```diff
@@ -27,35 +28,83 @@
...
+auto DetectionSensorBase::isWithinRange(
+  const geometry_msgs::Point & point1, const geometry_msgs::Point & point2,
+  const double range) const -> bool
+{
+  auto distanceX = point1.x() - point2.x();
+  auto distanceY = point1.y() - point2.y();
+  auto distanceZ = point1.z() - point2.z();
+
+  double distance = std::hypot(distanceX, distanceY, distanceZ);
+  return distance <= range;
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/src/sensor_simulation/detection_sensor/detection_sensor.cpp:40]:
```suggestion
  return std::hypot(point1.x() - point2.x(), point1.y() - point2.y(), point1.z() - point2.z()) <=
         range;
```

**Review by hakuturu583 — APPROVED** [review]

File: simulation/simple_sensor_simulator/src/sensor_simulation/detection_sensor/detection_sensor.cpp (hunk 4 above)
Comment (hakuturu583) [reply simulation/simple_sensor_simulator/src/sensor_simulation/detection_sensor/detection_sensor.cpp:40]: This comment about very small code styling, so I approve this PR.

**Pull request merged**